src/code42cli/cmds/alerts.py

```diff
@@ -180,6 +188,9 @@ def search(
         extractor.extract(*cli_state.search_filters)
     if handlers.TOTAL_EVENTS == 0 and not errors.ERRORED:
         echo("No results found.")
+    else:
+        output = _process_events(format, include_all, handlers.EVENTS)
```

Comment:
Shouldn't this happen in the handlers?

Reply:
I did initiate a [discussion](https://github.com/code42/code42cli/pull/133#discussion_r466994317) ,  I wanted processing of events outside the extraction module because it gives us flexibility to process data in their respective module. 

`security-data` doesn't need much processing as all the items are at top level data, whereas `alerts` can be processed as below in `alerts` module. 

```
def _transform_nested_events(events):
    transformed_events = []
    for event in events:
        transformed_event = {key: event[key] for key in event.keys() if type(event[key]) == str}
        
        for observation in event["observations"]:
            for key in observation["data"]:
                transformed_event["observation_"+key] = observation["data"][key]
        transformed_events.append(transformed_event)
    return transformed_events


def _process_events(output_format, include_all, events):
    if not include_all:
        format_header = get_format_header(include_all, events[0])
        return output_format(events, format_header)
    
    transformed_events = _transform_nested_events(events)
    format_header = {key: key.capitalize() for key in transformed_events[0].keys()}
    return output_format(transformed_events, format_header)

```

Reply:
I.e in `table` format and `csv` format we can display data by removing the nested structure.

Reply:
I was referring to above code in the meeting, replace `_process_events` method with above code and observe the difference in output with `--include-all` option. 

Also this is just an idea, we can polish it further if team has better ideas. 

All nested data keys are prepended with "observation_", that's because some keys from top level data  like `type` etc do not get overwritten.

Reply:
Perhaps, could you make them have / implement separate handlers instead?  

I am not a fan of holding onto `EVENTS`. They should be processed as they come through. Storing them in a class property to process later seems less functional and more of a work-around.

This is not a complete blocker for me since this all behind the scenes stuff ... but if we are going to have event handlers, they should handle the event

Reply:
I agree that `security-data` and `alerts` should have separate methods for processing data, but I think that should be handled by using separate handler instances (each with their own process logic defined) instead of moving the logic to be completely outside the handlers.

Reply:
also, since `observations` is a nested object and _especially_ because it could contain more than one item, I thought we had agreed that it should be excluded from being output in `alerts`, am I misunderstanding something?

Reply:
I also agree that the handlers should be defined separately, and we definitely want to process events as they come in, otherwise for large queries that might make multiple 10k search requests, the user would have to wait until they _all_ were retrieved before anything got output to the screen, which isn't a great experience.

Reply:
I agree with processing events as they come in and will get those changes done. 

I might have misread if [this](https://github.com/code42/code42cli/pull/133/files#r468700924) was finalised. So to clarify on it, current implementation is, we display all top level data by default. Now, if we do not display the nested object (`observations`) then do we need `--include-all` option!?

Reply:
The intention was not to output _all_ top-level fields by default, but rather to choose a _default_ set of top-level fields to output. Using `--include-all` would output all _top-level_ fields, but nested fields, lists, etc would never be output to the table.

Reply:
Changed as discussed.